src/sys/unix/uds.rs

```diff
@@ -0,0 +1,298 @@
+use crate::sys::unix::net::new_socket;
+
+use std::ascii;
+use std::cmp::Ordering;
+use std::ffi::OsStr;
+use std::fmt;
+use std::io;
+use std::mem;
+use std::os::unix::net::{UnixDatagram, UnixListener, UnixStream};
+use std::os::unix::prelude::*;
+use std::path::Path;
+
+/// An address associated with a mio specific unix socket.
+///
+/// This is implemented instead of imported from [`net::SocketAddr`] because
+/// there is no way to create a [`net::SocketAddr`]. One must be returned by
+/// [`accept`], so this is returned instead.
+///
+/// [`net::SocketAddr`]: std::os::unix::net::SocketAddr
+/// [`accept`]: #method.accept
+pub struct SocketAddr {
+    addr: libc::sockaddr_un,
+    len: libc::socklen_t,
+}
+
+enum AddressKind<'a> {
+    Unnamed,
+    Pathname(&'a Path),
+    Abstract(&'a [u8]),
+}
+
+pub fn connect_stream(path: &Path) -> io::Result<UnixStream> {
+    let socket = new_socket(libc::AF_UNIX, libc::SOCK_STREAM)?;
+    let (raw_addr, raw_addr_length) = socket_addr(path)?;
+    let raw_addr = &raw_addr as *const _ as *const _;
+
+    match syscall!(connect(socket, raw_addr, raw_addr_length)) {
+        Ok(_) => {}
+        Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {}
+        Err(e) => {
+            // Close the socket if we hit an error, ignoring the error
+            // from closing since we can't pass back two errors.
+            let _ = unsafe { libc::close(socket) };
+
+            return Err(e);
+        }
+    }
+
+    Ok(unsafe { UnixStream::from_raw_fd(socket) })
+}
+
+pub fn pair_stream() -> io::Result<(UnixStream, UnixStream)> {
+    let fds = [0, 0];
+    let flags = libc::SOCK_STREAM;
+
+    pair_descriptors(fds, flags)?;
+
+    Ok(unsafe {
+        (
+            UnixStream::from_raw_fd(fds[0]),
+            UnixStream::from_raw_fd(fds[1]),
+        )
+    })
+}
+
+pub fn bind_datagram(path: &Path) -> io::Result<UnixDatagram> {
+    let socket = new_socket(libc::AF_UNIX, libc::SOCK_DGRAM)?;
+    let (raw_addr, raw_addr_length) = socket_addr(path)?;
+    let raw_addr = &raw_addr as *const _ as *const _;
+
+    syscall!(bind(socket, raw_addr, raw_addr_length))?;
+    Ok(unsafe { UnixDatagram::from_raw_fd(socket) })
+}
+
+// pub fn accept(listener: &UnixListener) -> io::Result<Option<(UnixStream, SocketAddr)>> {
+pub fn accept(listener: &UnixListener) -> io::Result<(UnixStream, SocketAddr)> {
+    let mut storage: libc::sockaddr_un = unsafe { mem::zeroed() };
+    storage.sun_family = libc::AF_UNIX as libc::sa_family_t;
+    let mut len = mem::size_of_val(&storage) as libc::socklen_t;
+    let raw_storage = &mut storage as *mut _ as *mut _;
+
+    #[cfg(not(any(
+        target_os = "ios",
+        target_os = "macos",
+        target_os = "netbsd",
+        target_os = "solaris"
+    )))]
+    let sock_addr = {
+        let flags = libc::SOCK_NONBLOCK | libc::SOCK_CLOEXEC;
+
+        match syscall!(accept4(listener.as_raw_fd(), raw_storage, &mut len, flags)) {
```

Comment:
nit: why not just 
```rust
let sock_addr = {
    let flags =...;
    syscall!(accept4(listener.as_raw_fd(), raw_storage, &mut len, flags))?
};
```
rather than the match?